devito/at_controller.py

```diff
@@ -0,0 +1,223 @@
+
+            if len(blocks) > 1:
+                for y in range(minimum, maximum):
+                    blocks[1] = y if org_block_sizes[1] else None
+
+                    if len(blocks) > 2:
+                        for z in range(minimum, maximum):
+                            blocks[2] = z if org_block_sizes[2] else None
+                            block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+                    else:
+                        block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+            else:
+                block_list = block_list + [list(blocks)] if blocks not in block_list else block_list
+
+        # runs function for each block_size
+        for block in block_list:
+            self.op.propagator.block_sizes = block
+            times.append((block, self.get_execution_time()))
+
+        logger.info("Auto tuning using brute force complete")
+
+        times = sorted(times, key=lambda element: element[1])  # sorts the list of tuples based on time
+        self._write_block_report(times)  # writes the report
+
+    def _estimate_run_time(self, minimum, maximum):
+        """
+        Estimates run time for auto tuning
+        :param minimum: int - minimum tune range
+        :param maximum: int - maximum tune range
+        """
+        timing_run = 0  # estimating run time
+        for i in range(0, 5):
+            logger.info('Estimating auto-tuning runtime...sample %d/5' % (i + 1))
+
+            blocks = []
+            for block in self.op.propagator.block_sizes:
+                blocks = blocks + [random.randrange(minimum, maximum)] if block else blocks + [None]
+
+            self.op.propagator.block_sizes = blocks
+            timing_run += self.get_execution_time()
+
+        logger.info("Estimated runtime: %f minutes." % float(
+            timing_run / 5 * ((maximum - minimum) * (maximum - minimum)) / 600))
+
+    def get_execution_time(self):
+        """
+        Runs and times the function
+        :return: time - how long it took to execute
+        """
+        self.op.propagator.run(self.op.get_args())
+        return self.op.propagator.timings['kernel']
+
+    def _write_block_report(self, times):
+        """
+        Writes auto tuning report for block sizes
+        :param times: sorted list - times with block sizes
+        :raises IOError: if fails to write report
+        """
+        try:
+            full_report_text = ["%s %f\n" % (str(block), timee) for block, timee in times]
+
+            # Cache blocking dimensions
+            cb_dims = [True if block else False for block in self.op.propagator.block_sizes]
+            cb_dims = str(cb_dims).replace(" ", '')
+            shape = str(self.op.shape).replace(" ", '')
+
+            # Writes all auto tuning information into full report
+            with open(path.join(self.report_dir, "%s_time_o_%s_spc_bo_%s_shp_%s_b_%s_report.txt" %
```

Comment:
well its not used anywhere else except this method. SO no reason to extract it as that file is just for reference. Not used in data retrieving